### DISCUSSION TITLE
How to detect client disconnects on StreamingHttpResponse (Server Side Events)?

### DISCUSSION THREAD

**Discussion by AlTosterino** [Kludex/uvicorn#2338] [Q&A] [answered]
Hey!

We want to add Server Side Events in our application, but currently the while True loop runs indefinitely and client disconnect is not causing the task to stop.

Reproducible example:

```python
@api.get("/events")
async def get_events(request: ASGIRequest) -> StreamingHttpResponse:
    async def generate_message() -> AsyncGenerator[str, None]:
        while True:
            await asyncio.sleep(1)
            yield 'data: {"test": "testvalue"}\n\n'
            print("Working...")

    return StreamingHttpResponse(
        generate_message(),
        content_type="text/event-stream",
        headers={
            "X-Accel-Buffering": "no",
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
        },
    )
```

Above example will print Working... until server is closed (even if client closes connection and is not receiving the data).

Any way to detect client disconnect to stop the while True loop?

I'm not entirely sure if it's `django-ninja` issue, `uvicorn` issue or my issue 😅 

django-ninja==0.22.2
django==4.2.13
uvicorn[standard]==0.29.0

**Comment by Kludex** [comment, marked as answer]
It's a `django-ninja` issue. Uvicorn sends the `http.disconnect` event to the application as expected.

**Comment by AlTosterino** [comment]
It's not `django-ninja` issue, it's purely `Django` 
On Django < 5 this issue unfortunately exists (I'm using django==4.2.13)

https://code.djangoproject.com/ticket/34752
